peps/pep-0718.rst

```diff
@@ -62,6 +67,23 @@ If function objects were subscriptable, however, a more specific type could be g
 
    reveal_type(factory[int](lambda x: "Hello World" * x))  # type is Foo[int]
 
+Undecidable Inference
+^^^^^^^^^^^^^^^^^^^^^
+
+There are even cases where subclass relations make type inference impossible. However,
+if you can specialise the function type checkers can infer a meaningful type.
+
+.. code-block:: python
+
+   def foo[T](x: Sequence[T] | T) -> list[T]: ...
+
+   reveal_type(foo[bytes](b"hello"))
```

Comment:
This is maybe a better fit for the discussion, but I find this example disturbing, because it seems I could also write `foo[int](b"hello")`, match the `Sequence` part of the Union, and now the type checker would infer that the function returns `list[bytes]` instead of `list[int]`. Isn't that inherently unsound?

Reply:
Hmm, I do see your point here but isn't the current situation still unsound because the choosing of the type from the type checker is still up to its logic?